gtest/src/mailbox.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{manager::ExtManager, CoreLog, Log, RunResult};
+use codec::Encode;
+use gear_core::message::{Dispatch, DispatchKind, StoredMessage};
+use gear_core::{
+    ids::{MessageId, ProgramId},
+    message::Payload,
+};
+use std::cell::RefCell;
+
+pub struct Mailbox<'a> {
+    manager_reference: &'a RefCell<ExtManager>,
+    program_id: ProgramId,
+}
+
+impl<'a> Mailbox<'a> {
+    pub(crate) fn new(
+        program_id: ProgramId,
+        manager_reference: &'a RefCell<ExtManager>,
+    ) -> Mailbox<'a> {
+        Mailbox {
+            program_id,
+            manager_reference,
+        }
+    }
+
+    pub fn contains(&self, log: Log) -> bool {
+        self.manager_reference
+            .borrow_mut()
+            .actor_to_mailbox
+            .get(&self.program_id)
+            .expect("No mailbox with such program id")
+            .iter()
+            .any(|message| log.eq(message))
+    }
+
+    pub fn take_message(&self, log: Log) -> MessageReplier {
+        let index = self
+            .manager_reference
+            .borrow_mut()
+            .actor_to_mailbox
+            .get(&self.program_id)
+            .expect("No mailbox with such program id")
+            .iter()
+            .position(|message| log.eq(message))
+            .expect("No message that satisfies log");
+
+        let taken_message = self
+            .manager_reference
+            .borrow_mut()
+            .take_message(&self.program_id, index);
+
+        MessageReplier::new(taken_message, self.manager_reference)
+    }
+
+    pub fn reply(&self, log: Log, payload: impl Encode, value: u128) -> RunResult {
+        self.take_message(log).reply(payload, value)
+    }
+
+    pub fn reply_bytes(&self, log: Log, raw_payload: impl AsRef<[u8]>, value: u128) -> RunResult {
+        self.take_message(log).reply_bytes(raw_payload, value)
+    }
+}
+
+pub struct MessageReplier<'a> {
+    log: CoreLog,
+    manager_reference: &'a RefCell<ExtManager>,
+}
+
+impl<'a> MessageReplier<'a> {
+    pub(crate) fn new(
+        message: StoredMessage,
+        manager_reference: &'a RefCell<ExtManager>,
+    ) -> MessageReplier<'a> {
+        MessageReplier {
+            log: CoreLog::from_stored_message(message),
```

Comment:
After implementing `From`:
```suggestion
            log: message.into(),
```